### ISSUE TITLE
$TRAVIS_COMMIT_RANGE empty on branch builds

### ISSUE THREAD

**Issue by jasonkarns** [travis-ci/travis-ci#5133]
Possibly related to #1719, #2666 or #2668

Our test script leverages `$TRAVIS_COMMIT_RANGE`. It works fine on PRs but not on branches.

This owned [branch 'bad-checksum'](https://github.com/OiNutter/node-build/tree/bad-checksum) was pushed, triggering [build 295](https://travis-ci.org/OiNutter/node-build/builds/91816954) that succeeded, but it should have failed.

The associated [pull request 86](https://github.com/OiNutter/node-build/pull/86) triggered [build 296](https://travis-ci.org/OiNutter/node-build/builds/91817133) which failed, as expected.

As demonstrated on the PR itself, the PR build failed but the push build succeeded. (They both should have failed.)

![image](https://cloud.githubusercontent.com/assets/119972/11242921/e8c799d2-8dd2-11e5-9d7b-d615b47d9542.png)

**Comment by jasonkarns** [comment]
I should also note, there were no rebases or history manipulation of any kind.

1. single commit on new branch
2. branch pushed
3. pr opened

**Comment by jasonkarns** [comment]
some more digging suggests that a branch with >1 commit *does* populate the commit range. But a push with only 1 commit does not.

**Comment by jasonkarns** [comment]
Any movement on this?

**Comment by BanzaiMan** [comment]
Sorry about the delay. We will look into it. Thanks for your patience.

**Comment by jasonkarns** [comment]
Thanks! Anything in particular I can do to help?

**Comment by keradus** [comment]
@BanzaiMan commented on Jan 7
any update perhaps?

**Comment by BanzaiMan** [comment]
After some investigation, I see that `$TRAVIS_COMMIT_RANGE` is empty for the push builds exactly when the build is triggered by the first commit pushed to the new branch.

Compare [this initial build on the `BanzaiMan-patch-2` branch](https://travis-ci.org/BanzaiMan/travis_production_test/builds/142859422#L136) with [the second build](https://travis-ci.org/BanzaiMan/travis_production_test/builds/142860585#L136-L137).

**Comment by BanzaiMan** [comment]
This, I believe, is based on the GitHub payload. On the initial build, there is no base commit, so we do not define `$TRAVIS_COMMIT_RANGE`.

**Comment by jasonkarns** [comment]
This seems unsatisfactory. If a branch with >1 commit has a base commit, I don't understand why a branch with exactly 1 commit would not. Even if the github payload doesn't include a base commit (which sounds like a bug to be opened against their API?), would it not be possible to derive the base commit the same way that github does when generating the payload? Having TRAVIS_COMMIT_RANGE available for some branches but not others makes the variable rather untrustworthy.

**Comment by BanzaiMan** [comment]
I am not sure this is a "bug" in GitHub API. It seems to me that `git` itself is indifferent when it comes to new branch creation. When I create a new branch `topic` off a commit on another branch, `topic` does not retain any information about where it came from. When `topic` is pushed to GitHub, then, all they can guarantee is that the commit exists, but not what "commit range" constitutes the change for this push event.

In this case, Travis can set the commit range to be 0000…NEW_COMMIT (or some other "pseudo range"), but is it useful?

**Comment by keradus** [comment]
I ofen use sth like this:
```
- if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then COMMIT_RANGE=$TRAVIS_COMMIT_RANGE; else COMMIT_RANGE="HEAD~..HEAD"; fi;
- COMMIT_SCA_FILES=`git diff --name-only --diff-filter=ACMRTUXB $COMMIT_RANGE`
- php-cs-fixer fix --config=.php_cs.dist -v --dry-run --using-cache=no --path-mode=intersection $COMMIT_SCA_FILES
```

I would say that `0000...NEW_COMMIT` is pointless unless `HEAD~` doesn't exist

**Comment by BanzaiMan** [comment]
`COMMIT_RANGE="HEAD~..HEAD"` would contain only one commit, right? The first push event could contain multiple commits.

**Comment by jasonkarns** [comment]
@BanzaiMan That doesn't account for why the initial push on a new branch with _more_ than one commit doesn't exhibit this problem. There's no difference in git between a branch with 1 commit and a branch with >1 commit, so why would github provide the base commit in the latter case but not the former? That's why I'm suggesting this is a bug on their end.

**Comment by BanzaiMan** [comment]
Hmmm. You are quite right about that (if multiple commits appear in the first push to a new branch, a commit range is set). I'll contact GitHub support tomorrow.

**Comment by jasonkarns** [comment]
Cool! That's something I don't have much visibility into. :/ good luck!

**Comment by keradus** [comment]
@BanzaiMan yes, indeed. it;s a partially working solution until it will be fixed on Travis side.

**Comment by jdub** [comment]
Just ran into this. Did you have any luck with GitHub?

**Comment by jdub** [comment]
Interesting possibly related issue: Even if there are multiple commits on a new branch, the first commit is not included in the commit range.

**Comment by therealplato** [comment]
Thanks stale-bot, I've just encountered my first failed build when I pushed a single commit branch, and it took a while to track down this ticket to figure out what happened. I agree with @jasonkarns that the behavior's unsatisfactory so I'm reopening it.

If it's not technically possible to determine the base from which this push's commits was branched, and `TRAVIS_COMMIT_RANGE` would otherwise be empty, I would prefer that Travis sets @keradus 's suggestion `export TRAVIS_COMMIT_RANGE="HEAD~..HEAD"`. Or `<default branch>..HEAD`.

I trust Travis to provide a sane build environment, giving up entirely here feels insane to me.

**Comment by treyssatvincent** [comment]
> I trust Travis to provide a sane build environment, giving up entirely here feels insane to me.


Still not fixed in 2022 😓
